samples/99_zil_interpreter/app/zil_routine.rb

```diff
@@ -0,0 +1,148 @@
+          raise FunctionError, @name.to_s + '() - OPTIONAL list argument ' + argument.to_s + ' must have two elements!' if argument.class == Syntax::List && argument.elements.length != 2
+          arg_data = quoted ? argument.element : argument
+          @arguments << {arg_type: arg_type, arg_data: arg_data, arg_quoted: quoted}
+        elsif arg_type == :AUX
+          raise FunctionError, @name.to_s + '() - AUX list argument ' + argument.to_s + ' must have two elements!' if argument.class == Syntax::List && argument.elements.length != 2
+          arg_data = argument
+          @arguments << {arg_type: arg_type, arg_data: arg_data, arg_quoted: quoted}
+        else
+          raise FunctionError, @name.to_s + '() - Unsupported signature at ' + argument.to_s + '!'
+        end
+      end
+    end
+
+    def indent(width) # debugging only, put spaces based on stack level
+      str = ""
+      width.times do str += "  " end
+      str
+    end
+
+    def call(args, zil_context)
+      log "[ROUTINE.INVOKE] " + indent(zil_context.call_stack.length) + @name.to_s + " " + args.to_s + " {" if @debug_flag
+
+      # make sure we got enough parameters to fulfil our requirements
+      raise FunctionError, "Not enough arguments for #{@name}! (required: #{@num_required_arguments}, received: #{args.length})" if args.length < @num_required_arguments
+
+      result = nil
+      num_processed = 0
+      begin
+        # Push current locals onto stack
+        zil_context.call_stack.push([@name, args]) # this will be convenient to track nesting (and possibly debug deep procedure calls)
+        zil_context.locals_stack.push zil_context.locals
+
+        # Create new stack
+        zil_context.locals = {}
+
+        # Load function argument values onto new stack
+        i = 0
+        @arguments.each do |argument|
+          log "[ROUTINE.INVOKE] " + indent(zil_context.call_stack.length) + argument.to_s if @debug_flag
+
+          if argument[:arg_type] == :REQUIRED
+            value = argument[:arg_quoted] == true ? args[i] : eval_zil(args[i], zil_context)
+            zil_context.locals[argument[:arg_data]] = value
+            log "[ROUTINE.INVOKE] " + indent(zil_context.call_stack.length) + "  " + value.to_s if @debug_flag
+            num_processed += 1
+          elsif argument[:arg_type] == :OPTIONAL
+            provided = i < args.length
+            arg_data = argument[:arg_data]
+            if arg_data.class == Syntax::List
+              if provided
+                # if provided, eval and add to stack as list-1
+                value = eval_zil(args[i], zil_context)
+                zil_context.locals[arg_data.elements[0]] = value
+                num_processed += 1
+              else
+                # if NOT provided, add to stack with evaled list-2
+                value = eval_zil(arg_data.elements[1], zil_context)
+                zil_context.locals[arg_data.elements[0]] = value
+                num_processed += 1
+              end

```

Comment:
How about something like this?
```suggestion
              arg_name, default_value_expression = arg_data.elements
              value = eval_zil(
                provided ? args[i] : default_value_expression,
                zil_context
              )
              zil_context.locals[arg_name] = value
```

Reply:
👍